shell/src/context_listener.rs

```diff
@@ -118,35 +126,77 @@ fn listen_protocol_events(
                 event_count += 1;
 
                 match &msg {
-                    ContextAction::Set { block_hash: Some(block_hash), .. }
-                    | ContextAction::Copy { block_hash: Some(block_hash), .. }
-                    | ContextAction::Delete { block_hash: Some(block_hash), .. }
-                    | ContextAction::RemoveRecord { block_hash: Some(block_hash), .. }
-                    | ContextAction::Mem { block_hash: Some(block_hash), .. }
+                    ContextAction::Set { block_hash: Some(block_hash), key, value, .. } => {
+                        let state = get_default(&mut blocks, block_hash.clone());
+                        state.insert(key.join("/"), Bucket::Exists(value.clone()));
+
+                        context_storage.put(&block_hash.clone(), &ContextRecordValue::new(msg))?;
+                    }
+                    ContextAction::Copy { block_hash: Some(block_hash), to_key: key, from_key, .. } => {
+                        let partial_state = get_default(&mut blocks, block_hash.clone());
+
+                        let from_key = from_key.join("/");
+                        let to_key = key.join("/");
+                        if let Some(value) = partial_state.get(&from_key) {
+                            let value = value.clone();
+                            state.insert(to_key, value);
+                        } else if let Some(value) = state.get(&from_key) {
+                            let value = value.clone();
+                            state.insert(to_key, value);
+                        } else {
+                            warn!(log, "Trying to copy from non-existent location"; "from" => format!("{:?}", from_key), "to" => format!("{:?}", to_key));
+                        }
+
+                        context_storage.put(&block_hash.clone(), &ContextRecordValue::new(msg))?;
+                    }
+                    ContextAction::Delete { block_hash: Some(block_hash), key, .. }
+                    | ContextAction::RemoveRecord { block_hash: Some(block_hash), key, .. } => {
+                        let state = get_default(&mut blocks, block_hash.clone());
+                        state.insert(key.join("/"), Bucket::Deleted);
+
+                        context_storage.put(&block_hash.clone(), &ContextRecordValue::new(msg))?;
+                    }
+                    ContextAction::Commit { new_context_hash, block_hash: Some(block_hash), .. } => {
+                        if let Some(block) = blocks.get(block_hash) {
+                            state.merge(block);
+                        } else {
+                            warn!(log, "Trying to commit non-existent block"; "block" => format!("{:?}", block_hash));
```

Comment:
use `HashType` to format block hash here

Reply:
Fixed in c97f893